flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/InternalBacklogAwareTimerServiceImplTest.java

```diff
@@ -0,0 +1,111 @@
+
+    @Test
+    void testTriggerEventTimeTimer() throws Exception {
+        KeyGroupRange testKeyGroupRange = new KeyGroupRange(0, 1);
+        List<Tuple2<Integer, Long>> timers = new ArrayList<>();
+        TestProcessingTimeService processingTimeService = new TestProcessingTimeService();
+        final HeapPriorityQueueSetFactory priorityQueueSetFactory =
+                new HeapPriorityQueueSetFactory(testKeyGroupRange, 1, 128);
+        TimerSerializer<Integer, String> timerSerializer =
+                new TimerSerializer<>(IntSerializer.INSTANCE, StringSerializer.INSTANCE);
+        final TestTriggerable<Integer, String> triggerable =
+                TestTriggerable.eventTimeTrigger(
+                        (timer) -> timers.add(Tuple2.of(timer.getKey(), timer.getTimestamp())));
+        final KeyGroupedInternalPriorityQueue<TimerHeapInternalTimer<Integer, String>>
+                eventTimersQueue =
+                        createTimerQueue(
+                                "eventTimersQueue", timerSerializer, priorityQueueSetFactory);
+        final KeyGroupedInternalPriorityQueue<TimerHeapInternalTimer<Integer, String>>
+                processingTimerQueue =
+                        createTimerQueue(
+                                "processingTimerQueue", timerSerializer, priorityQueueSetFactory);
+        final BacklogTimeService<Integer, String> backlogTimeService =
+                createBacklogTimerService(
+                        processingTimeService, triggerable, processingTimerQueue, eventTimersQueue);
+        final TestKeyContext keyContext = new TestKeyContext();
+        final InternalTimerServiceImpl<Integer, String> internalTimerService =
+                createInternalTimerService(
+                        testKeyGroupRange,
+                        keyContext,
+                        processingTimeService,
+                        processingTimerQueue,
+                        eventTimersQueue);
+        internalTimerService.startTimerService(
+                IntSerializer.INSTANCE, StringSerializer.INSTANCE, triggerable);
+
+        final InternalBacklogAwareTimerServiceImpl<Integer, String> timerService =
+                new InternalBacklogAwareTimerServiceImpl<>(
+                        internalTimerService, backlogTimeService);
+
+        keyContext.setCurrentKey(1);
+        timerService.registerEventTimeTimer("a", 2);
+        timerService.registerEventTimeTimer("a", 1);
+        timerService.registerEventTimeTimer("a", 3);
+        keyContext.setCurrentKey(2);
+        timerService.registerEventTimeTimer("a", 3);
+        timerService.registerEventTimeTimer("a", 1);
+        timerService.advanceWatermark(2);
+        assertThat(timers).containsExactly(Tuple2.of(1, 1L), Tuple2.of(2, 1L), Tuple2.of(1, 2L));
+        timers.clear();
+
+        // switch to backlog processing
+        timerService.setBacklog(true);
+        timerService.setMaxWatermarkDuringBacklog(5);
```

Comment:
`AbstractStreamOperator.processRecordAttributes` would set watermark before set backlog, while the order in this test is reversed. It might be better to keep the same order as in production code.